sound/soc/sof/intel/hda-dsp.c

```diff
@@ -431,11 +435,60 @@ static int hda_dsp_set_D0_state(struct snd_sof_dev *sdev,
 		    !hda_enable_trace_D0I3_S0 ||
 		    sdev->system_suspend_target != SOF_SUSPEND_NONE)
 			flags = HDA_PM_NO_DMA_TRACE;
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
+		/*
+		 * Before system suspend CORB/RIRB DMA must be stopped and links powered off.
+		 * But when the system is in S0, this should be done only when there are no
+		 * codecs powered in order to save power.
+		 */
+		if (sdev->system_suspend_target == SOF_SUSPEND_NONE && bus->codec_powered)
```

Comment:
question: is it possible that the value of bus->codec_powered changes while the device is in D0ix?

Reply:
Yes, it is quite possible that codec_powered goes from powered down to up or the other way around when the DSP is in D0i3. Maybe we should power up links and start corb/rirb unconditionally when exiting D0i3?

But I wonder what happens if the codec is powered up when the DSP is in D0i3 due to jack detection. Will it fail because CORB/RIRB is stopped?

Reply:
This is kind of hard to test, Pierre. On our HDA platforms, we dont have WoV so the DSP will never be in D0i3 in practice.

Reply:
maybe we can look for the codec_mask and only do this for HDMI (mask 0x4). That way we don't take any chances with HDaudio codecs which we can't test anyways.

Reply:
ok sounds good

Reply:
done @plbossart

Reply:
@ranj063 just for my information - you mean we don't do WoV on HDA interfaces? So just using a topology like sof-hda-generic-kwd.m4 or sof-apl-keyword-detect.m4 wouldn't be enough for these tests?

Reply:
yeah we dont do WoV on any devices that have HDA codecs other than iDisp. So, the DSP would never enter D0i3. And we turn off the corb/rirb and links when the DSP enters D3 already.